### PR TITLE
tecmo_mix: Fixed the saturated colors in Final Star Force (Level 2)

### PR DESCRIPTION
**Pull request by darq-**

### CODE REVIEW

File: src/mame/video/tecmo_mix.cpp [hunk 1]
```diff
@@ -89,6 +89,28 @@ void tecmo_mix_device::set_bgpen(device_t &device, int bgpen)
...
+	uint8_t c2_g = c2 >> 8;
+	uint8_t c2_b = c2;
+
+	uint8_t a = std::min (0xFF, c1_a + c2_a);
+	uint8_t r = std::min (0xFF, c1_r + c2_r);
+	uint8_t g = std::min (0xFF, c1_g + c2_g);
+	uint8_t b = std::min (0xFF, c1_b + c2_b);
+
+	return ((a << 24) | (r << 16) | (g << 8) | b);
+}
```
Comment (cuavas) [inline src/mame/video/tecmo_mix.cpp]: Could you please change the following:
* Don't use gnu style, we prefer Allaman or K&R (i.e. use `func(arg)` and `arr[idx]` rather than `func (arg)` and `arr [idx]`)
* Make all your intermediates `const` if you're not planning to change them
* Don't play fast and loose with integers - you have all kinds of problems with overflow and abuse of `std::min`

The code definitely doesn't work as intended, the `std::min` calls are effectively doing nothing.  The second argument is the sum of two `uint8_t` which will not be promoted, so they'll exhibit modulo maths and the result of the addition will always be less than `0xff`.  You need to either make the intermediates larger (making them `unsigned int` is sensible, as it's going to be a type that's efficient to use), or explicitly cast one or both to something larger when adding.

Also, `0xff` is an `int` literal while the sum of two `uint8_t` is a `uint8_t` so the overload of `std::min` is ambiguous.

**Comment by darq-** [comment]
D'oh! I just noticed I clicked "Submit review" rather than "add a comment" when I wanted to leave a message.
I'm closing the wrong review.

**Comment by darq-** [comment]
Anyway, is it OK now?

**Comment by cuavas** [comment]
For consistency with the rest of the codebase - although there's a mix of Allman and K&R, we have pretty much no GNU-style formatting.  It makes a codebase harder to understand if it's inconsistent.

File: src/mame/video/tecmo_mix.h [hunk 2]
```diff
@@ -46,9 +46,7 @@ class tecmo_mix_device : public device_t,
 	int m_bgpen;
 
 private:
-
-
-
+	uint32_t sum_colors (const pen_t *pal, int c1_idx, int c2_idx);
```
Comment (cuavas) [inline src/mame/video/tecmo_mix.h]: GNU style - please make consistent with the rest of the code.

**Comment by darq-** [comment]
Sorry, I didn't understand, what are you asking for? I modified what you told me...

File: src/mame/video/tecmo_mix.cpp [hunk 3]
```diff
@@ -261,14 +275,13 @@ void tecmo_mix_device::mix_bitmaps(screen_device &screen, bitmap_rgb32 &bitmap,
 					if (fgbln)
 					{
 						// needs if bgpixel & 0xf check?
-						dd[x] = paldata[fgpixel + m_fgblend_source] + paldata[bgpixel + m_bgblend_comp];
+						dd[x] = sum_colors (paldata, fgpixel + m_fgblend_source, bgpixel + m_bgblend_comp);
```
Comment (cuavas) [inline src/mame/video/tecmo_mix.cpp]: GNU style - please make consistent with the rest of the code.

File: src/mame/video/tecmo_mix.cpp [hunk 4]
```diff
@@ -218,18 +235,15 @@ void tecmo_mix_device::mix_bitmaps(screen_device &screen, bitmap_rgb32 &bitmap,
...
-								dd[x] = paldata[fgpixel + m_fgblend_comp] + paldata[sprpixel + m_spblend_source];
+								dd[x] = sum_colors (paldata, fgpixel + m_fgblend_comp, sprpixel + m_spblend_source);
 							}
 						}
 						else // needs if bgpixel & 0xf check?
 						{
 							// blended sprite over solid bg pixel
-							dd[x] = paldata[bgpixel + m_bgblend_comp] + paldata[sprpixel + m_spblend_source];
-						//  dd[x] =  rand();
+							dd[x] = sum_colors (paldata, bgpixel + m_bgblend_comp, sprpixel + m_spblend_source);
```
Comment (cuavas) [inline src/mame/video/tecmo_mix.cpp]: GNU style - please make consistent with the rest of the code.

File: src/mame/video/tecmo_mix.cpp [hunk 5]
```diff
@@ -218,18 +235,15 @@ void tecmo_mix_device::mix_bitmaps(screen_device &screen, bitmap_rgb32 &bitmap,
 							else
 							{
 								// blended sprite over solid fgpixel?
-								dd[x] = paldata[fgpixel + m_fgblend_comp] + paldata[sprpixel + m_spblend_source];
+								dd[x] = sum_colors (paldata, fgpixel + m_fgblend_comp, sprpixel + m_spblend_source);
```
Comment (cuavas) [inline src/mame/video/tecmo_mix.cpp]: GNU style - please make consistent with the rest of the code.

File: src/mame/video/tecmo_mix.cpp [hunk 6]
```diff
@@ -193,16 +212,14 @@ void tecmo_mix_device::mix_bitmaps(screen_device &screen, bitmap_rgb32 &bitmap,
 							// needs if bgpixel & 0xf check?
 
 							//fg isn't used, sprite is used and blended with bg? -- used on trail of ball / flippers (looks odd)  -- some ninja gaiden enemy deaths (when behind fg) (looks ok?)  (maybe we need to check for colour saturation?)
-							dd[x] = paldata[bgpixel + m_bgblend_comp] + paldata[sprpixel + m_spblend_source];
+							dd[x] = sum_colors (paldata, bgpixel + m_bgblend_comp, sprpixel + m_spblend_source);
```
Comment (cuavas) [inline src/mame/video/tecmo_mix.cpp]: GNU style - please make consistent with the rest of the code.

File: src/mame/video/tecmo_mix.cpp [hunk 7]
```diff
@@ -171,20 +191,19 @@ void tecmo_mix_device::mix_bitmaps(screen_device &screen, bitmap_rgb32 &bitmap,
...
 
 								// fg is used and blended with sprite, sprite is used and blended with bg?  -- used on 'trail' of ball when ball is under the transparent area
-								dd[x] = paldata[bgpixel + m_bgblend_comp] + paldata[sprpixel + m_spblend_source]; // WRONG??
+								dd[x] = sum_colors (paldata, bgpixel + m_bgblend_comp, sprpixel + m_spblend_source); // WRONG??
 							}
 							else
 							{
 								// fg is used and blended with opaque sprite
-								dd[x] = paldata[fgpixel + m_fgblend_source] + paldata[sprpixel + m_spblend_comp];
+								dd[x] = sum_colors (paldata, fgpixel + m_fgblend_source, sprpixel + m_spblend_comp);
```
Comment (cuavas) [inline src/mame/video/tecmo_mix.cpp]: GNU style - please make consistent with the rest of the code.

File: src/mame/video/tecmo_mix.cpp [hunk 8]
```diff
@@ -171,20 +191,19 @@ void tecmo_mix_device::mix_bitmaps(screen_device &screen, bitmap_rgb32 &bitmap,
 								// needs if bgpixel & 0xf check?
 
 								// fg is used and blended with sprite, sprite is used and blended with bg?  -- used on 'trail' of ball when ball is under the transparent area
-								dd[x] = paldata[bgpixel + m_bgblend_comp] + paldata[sprpixel + m_spblend_source]; // WRONG??
+								dd[x] = sum_colors (paldata, bgpixel + m_bgblend_comp, sprpixel + m_spblend_source); // WRONG??
```
Comment (cuavas) [inline src/mame/video/tecmo_mix.cpp]: GNU style - please make consistent with the rest of the code.

File: src/mame/video/tecmo_mix.cpp [hunk 9]
```diff
@@ -89,6 +89,28 @@ void tecmo_mix_device::set_bgpen(device_t &device, int bgpen)
 	dev.m_bgpen = bgpen;
 }
 
+uint32_t tecmo_mix_device::sum_colors (const pen_t *pal, int c1_idx, int c2_idx)
```
Comment (cuavas) [inline src/mame/video/tecmo_mix.cpp]: GNU style - please make consistent with the rest of the code.

**Comment by darq-** [comment]
:/ ...Ok, I re-write the first question I asked (you didn't answer me)...

Are these change really essential to avoid a merge of a nice (I think) fix?
Do whitespaces cause any kind of bug?

Sorry but, I prefer to "waste" my time fixing bug / improving games rather than removing whitespaces...

Instead, if you find bad mistakes in my code (as you already did), I will happy to fix it and to follow your good advices.

Hope you can understand this.

**Comment by cuavas** [comment]
Yes they really are.
Refer to the MAME Coding Conventions: http://wiki.mamedev.org/index.php/MAME_Coding_Conventions
And also the coding standard notes in the readme: https://github.com/mamedev/mame/blob/master/README.md

Both contain the wording "Some parts of the code follow Allman style; some parts of the code follow K&R style -- mostly depending on who wrote the original version. Above all else, be consistent with what you modify, and keep whitespace changes to a minimum when modifying existing source. For new code, the majority tends to prefer Allman style, so if you don't care much, use that."

**Comment by cuavas** [comment]
Putting GNU-style code in the middle of an Allman-style file definitely violates "be consistent with what you modify"

**Comment by darq-** [comment]
I hope you was kidding when you said: "Yes they really are", again the most important think is the code correctness!

I just read those docs (quick look, though) and to be 'nitpicky', all code I seen already is a 'mix'..

For example:
Why "if (..." instead of "if(..." ?,
Why "for (int i..." instead of "for(int i..." ?,
Why "while (...." and not "while("...
...and so on...

If you don't want to merge because these stupid things... please close the PR...
I will not close it because it's your responsability to not merge an useful fix.

**Comment by happppp** [comment]
keywords aren't functions :)

**Comment by darq-** [comment]
Hi hap, nice to 'see' you again :)

Yes, of course :D, but what's the point? (they both have the 'same form')
Why "if (condition)" is good, instead "method (param);" is bad ?

**Comment by LnmVolbo** [comment]
@darq
The **point** is someone will have to review this code in the future and it's needlessly complicated to do so when the coding style is mixed.

I urge you to consider removing the unwanted whitespace so that the rest of the team members will not have such a difficult time in the future.

You would be completely correct if you were the only one working on this project.  But this is a _team_ effort and if you are unwilling to comply to the mutual benefit of the team, you must ask yourself why did you go to the trouble to submit a contribution?

This is not an opportunity to see how diverse the coding style can be in the codebase.  This is a chance to contribute to a project for the mutual benefit of all.

**Comment by happppp** [comment]
I don't think it's that bad, if this is your prefered code-style. The problem in this case is different styling in the same source file. Keep it consistent with the surrounding code.

**Comment by cuavas** [comment]
Allman and K&R styles both use spaces between flow control keywords and their associated expressions and in `if (x)`, `switch (x)` etc. but not function names and the opening parenthesis of their parameter/argument list, e.g. `void func(type param)` or `func(arg)`.  Functions are not the same thing as flow control constructs.  Whether you like GNU style better or not, MAME uses Allman and K&R, and we've documented that in multiple places.

As Abelson said, "Programs must be written for people to read, and only incidentally for machines to execute." (Structure and Interpretation of Computer Programs, 1984)  Mixing styles in the same unit makes the code messy, harder to read, and grating.  The time to clean this stuff up is before a pull request is merged, not afterwards, and if it's left until afterwards, it will be a member of the team that has to clean it up.  We have enough mess in out codebase already, we're trying to maintain some basic standards for new code.

**Comment by darq-** [comment]
@LnmVolbo:
Would you have some difficulties to read the code because **one** additional whitespace? :O
I could understand you, if my code was all in a single line...

I tried to contribute because, I thought improving a game was a good think...

I had having fun to write code, but if I have to be annoyed with these 	irrelevant thinks, 'the fun' will fade out suddendly...

Additionally, did you realize what a waste of time is this 'thread' ?
I was writing other nice and major fixes for another game... but I stopped to write because this discussion...

@happppp:
The point is that is not my style... again i don't see any style diff. to write a method or a keyword...

**Comment by darq-** [comment]
@cuavas:
Ah, ok, too bad then.

I already wrote it above... I can't believe that a code with one whitespace is difficult to read...
Again, you're free to close this PR... but of course, if anybody cares, my collaboration will ends here.

To all:
Sorry for my english... I know it sucks :)

**Comment by LnmVolbo** [comment]
@darq,

Of course improving a game is a "good think".

But making the sourcecode more difficult to maintain is a "bad think".

Haven't you ever found some old sourcecode and needed to improve it, and found that it was in such a strange coding style that it made the process much more difficult than it needed to be?

Even though improving a game is good, improving the game at the expense of the overall project is bad.

And of course it is important to keep it fun.  But there are more than the two alternatives than expressed so far:

1. using bad style, but fun

2. not contributing any more because it is not fun.

There is a third alternative.  You can ask another contributor here to help you with the style changes or make the style changes for you.  You might get a little flak, but you might still have fun.

This is a team project.  There are resources here that can help you.  All you have to do is show some flexibility and ask for help.

**Comment by darq** [comment]
Hi, @LnmVolbo,  thanks for mentioning @darq instead of @darq- it was great to see your condescending tone in my inbox.  Please be more careful in the future, and exclude me from future discussions.

**Comment by darq-** [comment]
`<ot>`
lol, too funny... 'two' darq here...
He 'stoled' my nick :D ...I had to add a "-" when I created my github account... :)
`</ot>`

@LnmVolbo :
Quick answer, I have to go to work...

Yes, in the past I found some ugly source codes and I understand what you said, but I see that you continue to insist that I used bad style...

The point here is not the style (also because I normally use 'MAME style' when I write my programs), but a single stupid whitespace, that among other things, would make even a better separation between method's name and parameters, just like keywords.

To quit or not to quit, depends on the cuavas' move... at the moment I am in an 'idle state' :D, but I'm inclined to pick the second option.
The third alternative is pointless, I don't need help to add or remove a single whitespace (a random finger can do the work ;))

**Comment by rb6502** [comment]
@darq- I gather you've never worked professionally on software?  This spacing stuff seems dumb, but it's actually pretty important.  Just spend the 5 or 10 keystrokes to fix it and we're all good :)

**Comment by darq-** [comment]
@rb6502 
Hi, R. Belmont :)

Actually I am a professional programmer, but never had those kinds of problems.
On the contrary if in a 'work environment' there was someone who 'stopped' a release or a feature just because these insignificant things, he would be fired immediately.

Anyway, you got the point, this spacing stuff **WAS** a very dumb think indeed, but it was more as a matter of principle than a real important think.

When cuavas told me to change whitespaces in my method, even though I didn't like, I changed them to make him happy :)
I thought it was over, but when he insisted to analyze each single character I wrote, then I didn't like this kind of attitude.

Moreover, I already told him, that I would happy to fix any valid error in my code and to follow any good suggestion...
I still remember when you told me on the mw forum, that using hold_line wasn't a good think :D, and in fact, when possible, I try to avoid them, and hunting for ack_w instead ;)


@Osso13:
Thank you for merging the PR, but you could avoid your second commit...
Anyway while we are here... you forgot to modify, for example, the following:

```
cuavas mode: on
nitpick_val <<= 3;
```
gaiden.cpp:
descramble_drgnbowl method -> there are 2 for with this form: "for( i" instead of "for (i"...
all PORT_MODIFY macros (but there are more) have 'attached' parenthesis, while the rest of macros have multiple whitespaces...

tecmo16.cpp:
Additional spaces in MCFG_TECMO_MIXER_BLENDCOLS, MCFG_TECMO_MIXER_REGULARCOLS and MCFG_TECMO_MIXER_BLENDSOURCE macros...
(compared to the nearby macros)

`cuavas mode: off`

...and I just made, only a quick check... :)

Seriously, this was just a silly example to highlight how stupid, annoying and a total waste of time, can be such kind of things...

To summarize:
If contributing (with few lines of code) has to be a war... thanks, but, I'm not willing to participate, otherwise (again, if anybody cares) I can continue to commit some fixes...

Please let me let me know.

**Comment by Osso13** [comment]
@darq- Disclaimer: I'm speaking only for myself and not for Mamedev as a group.
I only addressed the coding style differences in your patch. I won't change what's already there. This is consistent with what Vas asked you to do, which is to adapt to the coding style of the file you touch for the lines you change, not to scan the entire file and update what was already there.
I hope that you'll find the motivation to continue to provide your fixes, which IMO are great. I also hope that you will understand that Vas is acting for the good of the project (it's his task and prerogative as a member of the MAME board) and he hasn't got anything personal against you.

**Comment by mamehaze** [comment]
politics and nonsense aside, thanks for looking at this, I wasn't actually aware there were still any issues with my implementation of the blending here, so I appreciate the work done on the code in not only identifying the problem but fixing it.

I also appreciate the approach Osso took, I think the aggression, pedantic comments and condescending attitudes in the initial comments were rather unnecessary when surely it would have taken much less time and effort to do that in the first place with none of this thread being neccessary (as in accept the valid and important emulation improvement, with the team taking care of any adjustments which are merely preference)  This has traditionally been the way of the project, so I'm not really sure why things have flared up here, so I'd like to apologise for that.

It would be a shame to lose your contributions over such a disagreement as the things you've been looking at are exactly the kind of thing that have needed looking at for years.

**Comment by darq-** [comment]
lol, you're welcome :)
...I saw you emulated almost the whole universe in MAME :), so your comment is really appreciated :D

Honestly, I'd like to continue, but first of all, I'd like to see a little bit of flexibility (1% would be sufficient)...

I can understand guidelines, but not as absolute law.
I didn't use a different code-style at all (for example inlined curly bracket)... my only "sin" (according to cuavas) was a single whitespace after the method's name.

**Comment by cuavas** [comment]
We are already far more flexible than most projects when it comes to formatting.  If you think our standards are onerous, you'd scream if you saw how prescriptive the guidelines are for Linux kernel, BSD KNF, or pretty much any other major open source project.  The extra space does constitute a different style - both Allman and K&R styles use no space between function names and opening parenthesis of the parameter/argument list for functions to emphasise the fact that that they are not flow control constructs.
